File: web/src/services/event.ts
import api from './api'
import {
  SecurityEvent,
  EventFilter,
  ApiResponse,
  PaginatedResponse,
} from '@/types'

// 后端 /event/v1/list 返回格式
interface EventListResponse {
  total: number
  events: Array<{
    id: string
    title: string
    severity: string
    source: string
    source_url?: string
    status: string
    cve_id?: string
    risk_score?: number
    created_at: string
  }>
}

// 后端 /event/v1/stats 返回格式
interface EventStatsResponse {
  total: number
  today_count: number
  critical_count: number
  by_severity: Record<string, number>
}

// 后端 /event/v1/trend 返回格式
interface EventTrendResponse {
  items: Array<{
    date: string
    total: number
    critical: number
    high: number
    medium: number
    low: number
  }>
}

export const eventService = {
  // 获取事件列表（后端 GET /event/v1/list?limit=&offset=&severity=&status=&keyword=&order_by=&order_dir=）
  async list(filter?: EventFilter): Promise<PaginatedResponse<SecurityEvent>> {
    const page = filter?.page || 1
    const size = filter?.size || 20
    const params: Record<string, unknown> = {
      limit: size,
      offset: (page - 1) * size,
    }
    if (filter?.severity) params.severity = filter.severity
    if (filter?.status) params.status = filter.status
    if (filter?.keyword) params.keyword = filter.keyword
    if (filter?.order_by) params.order_by = filter.order_by
    if (filter?.order_dir) params.order_dir = filter.order_dir

    const res = await api.get<ApiResponse<EventListResponse>>('/event/v1/list', { params })
    const data = res.data.data
    const events = data?.events || []

    const list: SecurityEvent[] = events.map(item => ({
      id: item.id,
      subscription_id: 0,
      title: item.title,
      severity: (item.severity || 'medium') as SecurityEvent['severity'],
      status: (item.status || 'new') as SecurityEvent['status'],
      source: item.source || '',
      source_url: item.source_url || '',
      event_time: item.created_at,
      cve_id: item.cve_id,
      cvss_score: item.risk_score,
      created_at: item.created_at,
    }))

    return {
      list,
      total: data?.total ?? list.length,
      page,
      size,
    }
  },

  // 获取单个事件（后端无此接口，从列表查找）
  async get(id: string): Promise<SecurityEvent> {
    const res = await this.list({ size: 100 })
    const found = res.list.find(e => e.id === id)
    if (!found) throw new Error('事件不存在')
    return found
  },

  // 更新事件状态（POST /event/v1/update_status）
  async updateStatus(id: string, status: string): Promise<void> {
    await api.post('/event/v1/update_status', { id, status })
  },

  // 删除安全事件（POST /event/v1/delete）
  async delete(id: string): Promise<void> {
    await api.post('/event/v1/delete', { id })
  },

  // 获取事件统计（调用后端 GET /event/v1/stats）
  async getStats(): Promise<{
    total: number
    today_count: number
    critical_count: number
    by_severity: Record<string, number>
    by_status: Record<string, number>
  }> {
    try {
      const res = await api.get<ApiResponse<EventStatsResponse>>('/event/v1/stats')
      const data = res.data.data
      return {
        total: data?.total || 0,
        today_count: data?.today_count || 0,
        critical_count: data?.critical_count || 0,
        by_severity: data?.by_severity || {},
        by_status: {},
      }
    } catch {
      return { total: 0, today_count: 0, critical_count: 0, by_severity: {}, by_status: {} }
    }
  },

  // 获取事件趋势（调用后端 GET /event/v1/trend?days=N）
  async getTrend(days = 30): Promise<Array<{
    date: string
    total: number
    critical: number
    high: number
    medium: number
    low: number
    info: number
  }>> {
    try {
      const res = await api.get<ApiResponse<EventTrendResponse>>('/event/v1/trend', { params: { days } })
      return (res.data.data?.items || []).map(item => ({ ...item, info: 0 }))
    } catch {
      return []
    }
  },

  // AI分析事件（后端无此接口，暂时 mock）
  async analyze(_id: string): Promise<{ risk_score: number; severity: string; recommendation: string }> {
    return { risk_score: 0, severity: 'medium', recommendation: '' }
  },

  // 多Agent流水线处理（后端无此接口，改用 pipeline stream）
  async processPipeline(): Promise<{ total_count: number; dedup_count: number; new_count: number; processed_at: string; steps: Array<{agent: string; status: string; message: string; count: number}> }> {
    // 后端无 /event/pipeline/process，返回空结果
    return {
      total_count: 0,
      dedup_count: 0,
      new_count: 0,
      processed_at: new Date().toISOString(),
      steps: [],
    }
  },
}

File: internal/controller/event/event.go
// Package event 提供安全事件 HTTP 控制器。
// 职责仅限 HTTP 层：解析请求 → 调用 eventsvc → 映射响应 DTO。
// 业务逻辑（severity 默认值、risk_score 自动推算）已下沉至 internal/service/event。
// PipelineStream 直接调用 AI 管道，不含 DB 操作，保留在控制器层。
package event

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"

	v1 "Fo-Sentinel-Agent/api/event/v1"
	"Fo-Sentinel-Agent/internal/ai/agent/event_analysis_pipeline"
	"Fo-Sentinel-Agent/internal/ai/models"
	"Fo-Sentinel-Agent/internal/ai/trace"
	eventsvc "Fo-Sentinel-Agent/internal/service/event"
	"Fo-Sentinel-Agent/utility/sse"

	"github.com/cloudwego/eino/schema"
	"github.com/gogf/gf/v2/frame/g"
)

type ControllerV1 struct{}

func NewV1() *ControllerV1 {
	return &ControllerV1{}
}

// List 返回事件列表，委托 eventsvc.List 查询，结果映射为 API DTO。
// 数据库不可用时返回空列表而非报错，保证前端不崩溃。
func (c *ControllerV1) List(ctx context.Context, req *v1.ListReq) (*v1.ListRes, error) {
	events, total, err := eventsvc.List(ctx, req.Limit, req.Offset, req.Severity, req.Status, req.Keyword, req.OrderBy, req.OrderDir)
	if err != nil {
		// 数据库不可用时优雅降级，返回空列表
		return &v1.ListRes{Total: 0, Events: []v1.EventItem{}}, nil
	}
	items := make([]v1.EventItem, 0, len(events))
	for _, e := range events {
		// 从 metadata JSON 中提取原始链接
		sourceURL := ""
		if e.Metadata != "" {
			var meta map[string]string
			if err2 := json.Unmarshal([]byte(e.Metadata), &meta); err2 == nil {
				sourceURL = meta["link"]
			}
		}
		items = append(items, v1.EventItem{
			ID:        e.ID,
			Title:     e.Title,
			Content:   "",
			EventType: e.EventType,
			Severity:  e.Severity,
			Source:    e.Source,
			SourceURL: sourceURL,
			Status:    e.Status,
			CVEID:     e.CVEID,
			RiskScore: e.RiskScore,
			CreatedAt: e.CreatedAt.Format("2006-01-02T15:04:05Z07:00"),
		})
	}
	return &v1.ListRes{Total: total, Events: items}, nil
}

// Stats 返回事件统计数据：总数、今日新增、高危数量、按严重程度分组、近7天新增、待处置数。
func (c *ControllerV1) Stats(ctx context.Context, req *v1.StatsReq) (*v1.StatsRes, error) {
	stats, err := eventsvc.Stats(ctx)
	if err != nil {
		return &v1.StatsRes{BySeverity: map[string]int64{}}, nil
	}
	return &v1.StatsRes{
		Total:         stats.Total,
		TodayCount:    stats.TodayCount,
		CriticalCount: stats.CriticalCount,
		BySeverity:    stats.BySeverity,
		New7Days:      stats.New7Days,
		Pending:       stats.Pending,
	}, nil
}

// Trend 返回最近 N 天的事件趋势，按日期分组，数据库不可用时返回空列表。
func (c *ControllerV1) Trend(ctx context.Context, req *v1.TrendReq) (*v1.TrendRes, error) {
	items, err := eventsvc.Trend(ctx, req.Days)
	if err != nil {
		return &v1.TrendRes{Items: []v1.TrendItem{}}, nil
	}
	result := make([]v1.TrendItem, 0, len(items))
	for _, it := range items {
		result = append(result, v1.TrendItem{
			Date:     it.Date,
			Total:    it.Total,
			Critical: it.Critical,
			High:     it.High,
			Medium:   it.Medium,
			Low:      it.Low,
		})
	}
	return &v1.TrendRes{Items: result}, nil
}

// Create 手动创建安全事件，委托 eventsvc.Create 处理默认值和入库。
func (c *ControllerV1) Create(ctx context.Context, req *v1.CreateReq) (*v1.CreateRes, error) {
	id, err := eventsvc.Create(ctx, req.Title, req.Content, req.Severity, req.Source, req.CVEID, req.RiskScore)
	if err != nil {
		return nil, err
	}
	return &v1.CreateRes{ID: id}, nil
}

// UpdateStatus 更新单条事件状态，委托 eventsvc.UpdateStatus 处理。
func (c *ControllerV1) UpdateStatus(ctx context.Context, req *v1.UpdateStatusReq) (*v1.UpdateStatusRes, error) {
	if err := eventsvc.UpdateStatus(ctx, req.ID, req.Status); err != nil {
		return nil, err
	}
	return &v1.UpdateStatusRes{}, nil
}

// Delete 软删除安全事件，委托 eventsvc.Delete 处理。
func (c *ControllerV1) Delete(ctx context.Context, req *v1.DeleteReq) (*v1.DeleteRes, error) {
	if err := eventsvc.Delete(ctx, req.ID); err != nil {
		return nil, err
	}
	return &v1.DeleteRes{}, nil
}

// BatchDelete 批量软删除安全事件，委托 eventsvc.BatchDelete 处理。
func (c *ControllerV1) BatchDelete(ctx context.Context, req *v1.BatchDeleteReq) (*v1.BatchDeleteRes, error) {
	if err := eventsvc.BatchDelete(ctx, req.IDs); err != nil {
		return nil, err
	}
	return &v1.BatchDeleteRes{}, nil
}

// BatchUpdateStatus 批量更新安全事件状态，委托 eventsvc.BatchUpdateStatus 处理。
func (c *ControllerV1) BatchUpdateStatus(ctx context.Context, req *v1.BatchUpdateStatusReq) (*v1.BatchUpdateStatusRes, error) {
	if err := eventsvc.BatchUpdateStatus(ctx, req.IDs, req.Status); err != nil {
		return nil, err
	}
	return &v1.BatchUpdateStatusRes{}, nil
}

// PipelineStream 触发 Event Analysis Agent（ReAct 智能体）进行流式事件分析，SSE 逐 chunk 推送结果。
func (c *ControllerV1) PipelineStream(ctx context.Context, req *v1.PipelineStreamReq) (*v1.PipelineStreamRes, error) {
	// 启动链路追踪（标记为独立事件分析，非会话内操作）
	tags := map[string]any{"context": "standalone_event_analysis"}
	ctx = trace.StartRun(ctx, "event.pipeline", "/api/event/v1/pipeline/stream", "", 0, req.Query, tags)
	var pipelineErr error
	defer func() { trace.FinishRun(ctx, pipelineErr) }()

	client := sse.NewClient(g.RequestFromCtx(ctx))

	// 获取事件分析 Agent 单例（内含 RAG 管道 + ReAct 推理循环）
	runner, err := event_analysis_pipeline.GetEventAnalysisAgent(ctx)
	if err != nil {
		pipelineErr = err
		client.Send("error", err.Error())
		client.Done()
		return nil, nil
	}

	// AGENT span：使内部 LLM/TOOL/RETRIEVER 节点归属到 EventAnalysisAgent 下，
	// 而非直接挂在链路根节点，保持与 intent/executor.go 路径的一致性。
	spanCtx, spanID := trace.StartSpan(ctx, trace.NodeTypeAgent, "EventAnalysisAgent")

	// 以流式模式调用 Agent，逐 chunk 接收推理过程和最终结论
	sr, err := runner.Stream(spanCtx, &event_analysis_pipeline.UserMessage{
		ID:      "stream",
		Query:   req.Query,
		History: nil, // 事件分析不携带历史上下文，每次独立分析
	})
	if err != nil {
		pipelineErr = err
		trace.FinishSpan(spanCtx, spanID, err, nil)
		client.Send("error", err.Error())
		client.Done()
		return nil, nil
	}
	defer sr.Close()

	var spanErr error
	for {
		chunk, err := sr.Recv()
		if errors.Is(err, io.EOF) {
			// 流式输出结束，通知前端关闭 SSE 连接
			client.Send("done", "Stream completed")
			break
		}
		if err != nil {
			pipelineErr = err
			spanErr = err
			client.Send("error", err.Error())
			break
		}
		// 过滤纯空白 chunk，避免无意义推送
		// 将内容包装为 JSON 格式，与前端解析协议对齐：{"type":"content","content":"..."}
		if chunk != nil && strings.TrimSpace(chunk.Content) != "" {
			jsonBytes, _ := json.Marshal(map[string]string{"type": "content", "content": chunk.Content})
			client.Send("message", string(jsonBytes))
		}
	}
	trace.FinishSpan(spanCtx, spanID, spanErr, nil)
	client.Done()
	return nil, nil
}

// AnalyzeSingleStream 针对单条安全事件直接调用 LLM，SSE 流式返回结构化 AI 解决方案。
// 直接调用模型（无 ReAct 工具调用），避免多步工具循环导致的延迟或步数耗尽问题，
// 响应更快，可靠性更高。
func (c *ControllerV1) AnalyzeSingleStream(ctx context.Context, req *v1.AnalyzeSingleStreamReq) (*v1.AnalyzeSingleStreamRes, error) {
	client := sse.NewClient(g.RequestFromCtx(ctx))

	g.Log().Infof(ctx, "[AnalyzeSingleStream] 收到请求 | event_id=%s | title=%s | severity=%s", req.EventID, req.Title, req.Severity)

	m, err := models.OpenAIForDeepSeekV3Quick(ctx)
	if err != nil {
		g.Log().Errorf(ctx, "[AnalyzeSingleStream] 模型初始化失败 | err=%v", err)
		client.Send("error", err.Error())
		client.Done()
		return nil, nil
	}

	cveInfo := ""
	if req.CVEID != "" {
		cveInfo = "CVE 编号：" + req.CVEID + "\n"
	}

	systemContent := `你是安全事件应急响应专家。针对用户提供的安全事件，基于安全知识和行业最佳实践，生成一份结构化的应急响应方案。

## 输出格式（严格 Markdown，必须包含以下三节）

### 应急响应步骤
- 立即需要采取的行动（分点列出，按优先级排序）

### 修复方案
- 具体的技术修复措施（版本升级、配置变更、补丁路径等）

### 长期防护建议
- 系统性安全加固措施（架构改进、监控策略、定期审计等）

输出使用中文，技术术语保留英文。`

	userContent := fmt.Sprintf(
		"针对以下安全事件，请提供详细的 AI 解决方案：\n\n事件标题：%s\n严重程度：%s\n%s数据来源：%s",
		req.Title, req.Severity, cveInfo, req.Source,
	)

	messages := []*schema.Message{
		schema.SystemMessage(systemContent),
		schema.UserMessage(userContent),
	}

	sr, err := m.Stream(ctx, messages)
	if err != nil {
		g.Log().Errorf(ctx, "[AnalyzeSingleStream] Stream 调用失败 | err=%v", err)
		client.Send("error", err.Error())
		client.Done()
		return nil, nil
	}
	defer sr.Close()

	chunkCount := 0
	for {
		chunk, err := sr.Recv()
		if errors.Is(err, io.EOF) {
			g.Log().Debugf(ctx, "[AnalyzeSingleStream] Stream 结束 | chunks=%d", chunkCount)
			client.Send("done", "Stream completed")
			break
		}
		if err != nil {
			// context.Canceled 是客户端主动断开连接（前端关闭面板），属正常行为，降级为 info 避免误报
			if errors.Is(err, context.Canceled) {
				g.Log().Warningf(ctx, "[AnalyzeSingleStream] 客户端主动断开连接 | chunks=%d", chunkCount)
			} else {
				g.Log().Errorf(ctx, "[AnalyzeSingleStream] Stream 接收异常 | chunks=%d | err=%v", chunkCount, err)
				client.Send("error", err.Error())
			}
			break
		}
		if chunk == nil {
			g.Log().Debugf(ctx, "[AnalyzeSingleStream] 收到空 chunk，跳过")
			continue
		}
		if strings.TrimSpace(chunk.Content) != "" {
			chunkCount++
			jsonBytes, _ := json.Marshal(map[string]string{"type": "content", "content": chunk.Content})
			client.Send("message", string(jsonBytes))
		}
	}
	client.Done()
	return nil, nil
}
